### ISSUE TITLE
perf(model): route Metal hybrid MTP through wide-M batched verification

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12349]
Parent: #12235
Follow-on to closed #12237
Related: #12238

## Classification
- **Centrality:** Core
- **Worker tier:** Sol implementation; Astra parity review if recurrence semantics change
- **Write set:** `internal/model/verify.go`, `internal/model/metal_mtp.go`, focused `internal/model/*test.go`

## For / Problem / Today / Better because / Witness
- **For:** Apple Silicon Qwen hybrid sessions expecting one-pass speculative verification.
- **Problem:** the production coordinator must reach the wide-M verifier for the actual Metal/quantized/hybrid envelope.
- **Today:** `verifyForwardBatchedOK` rejects Metal, backend, quantized, and hybrid sessions, so `MetalMTPCoordinator` falls back to a serial `Step` loop.
- **Better because:** accepted wide-M kernels become reachable in the product path, allowing one weight-streaming pass while retaining exact serial parity.
- **Witness:** `TestMetalMTPWideMUsesBatchedVerifier`.

## Acceptance
- Admit only an explicitly supported Metal/Qwen hybrid envelope.
- Prove batched dispatch is selected and its logits/state equal serial verification.
- Keep unsupported shapes on typed fak-native serial decode; never silently switch engines.

## Witness command
`go test ./internal/model -run '^TestMetalMTPWideMUsesBatchedVerifier$' -count=1`

## Non-goals
New shader kernels, new quant formats, benchmark claims, or universal batched admission.

## Dispatch contract repair (2026-09-12)

```routing
lane: model
paths: ["internal/model/verify.go", "internal/model/metal_mtp.go", "internal/model/metal_mtp_test.go"]
expected_steps: 4
```

## Parent context
#12235; completed predecessor #12237; Coordinates with: #12238

## Current state
`verifyForwardBatchedOK` rejects the supported Metal/quantized/Qwen-hybrid tuple, so the production coordinator falls back to serial `Step` verification.

## Why this is next
The accepted wide-M implementation remains unreachable from the native production coordinator until this narrow admission path is proven parity-safe.

## Working spine
Capture route selection, admit only the supported tuple, and compare batched logits and state with serial verification.

## Scope
The first executable slice owns the full issue outcome in the routed files: supported-envelope admission, batched-dispatch selection, exact serial parity, and typed fak-native fallback elsewhere.

## Gold-plating boundary
No shader kernel, quant format, benchmark claim, or universal batched admission.

## Done condition
The supported Metal Qwen hybrid tuple selects wide-M verification with serial-equivalent logits and state; unsupported tuples stay serial.

## Witness
`go test ./internal/model -run '^TestMetalMTPWideMUsesBatchedVerifier$' -count=1`

## Acceptance gate
The focused route/parity regression passes and owned-path validation succeeds; performance promotion remains separately hardware-witnessed.

## Lane
model

## Closure binding
The resolving explicit-path commit cites #12349 and carries `(fak model)`; closed #12237 supplies the predecessor and #12238 is nonblocking coordination.

**Comment by anthony-chaudhary** [comment]
Status/Verdict: SW-VERIFIED (PENDING_HARDWARE)
Changed files: internal/model/verify.go, internal/model/metal_mtp.go, internal/model/metal_mtp_test.go | Landed commit: 0cb93174cca505cee6dc44cb497eb45d3e09281b on origin/main (tip f8133b00b89a833be42e42ab0c0ecd0e47af7cd3)
Test output: TestMetalMTPWideMUsesBatchedVerifier (4/4 subtests PASS, 0.01s), TestMetalMTP* suites PASS | Next physical command: go test -v ./internal/model -run '^TestMetalMTPWideMUsesBatchedVerifier$' -count=1

**Comment by anthony-chaudhary** [comment]
Status/Verdict: VERIFIED (PASS)
Changed files: internal/model/verify.go, internal/model/metal_mtp.go, internal/model/metal_mtp_test.go | Landed commit: 0cb93174cca505cee6dc44cb497eb45d3e09281b on origin/main (tip 9a6e28e2328e)
Test output: TestMetalMTPWideMUsesBatchedVerifier (4/4 PASS, 0.00s), TestMetalMTP* (18/18 PASS, 6.64s), go vet clean

**Comment by anthony-chaudhary** [comment]
Parent acceptance remains RED. Child #12914 has independently raw-checked physical evidence that narrows the #12853 diagnosis: https://github.com/anthony-chaudhary/fak/issues/12914#issuecomment-5647991601

On the exact native Apple M3 Pro P4/H256 fixture, all1024 quantized codes match and all four normalized rows are exact, while 3/32 scale bit patterns differ (first host3cd63ec4/device3cd63ec3, row1/token5/block5). H/D/S projected output hashes remain unchanged and pairwise unequal; H-versus-S under fixed host quantization is a separate route discrepancy. These observations do not establish a sole cause or satisfy original parity.

Actual executor56092 exit1, 1 test/0 skips; strict logits/GDN/KV/hidden/continuation failures retained. Lifecycle checks and separate1152-byte diagnostic accounting succeeded. All1702 frozen source entries unchanged; owned group1366/descendants absent and canonical lease released. No commit, landing, retry or new production edit.

Preserved receipt SHA2561a33c28ca1e0cdda0fef5709d11b2d64a10490968d0548f2b6008b928e59c598 in managed model-165612fe8f4e/_scratch/issue-12914-hil. Full raw arrays were not logged; every element was compared and hashes/counts/first difference captured.

Next: independently review a minimal scale-formation candidate while retaining the separate H-versus-S investigation and original full parent acceptance gates.

**Comment by anthony-chaudhary** [comment]
One physical #12919 candidate witness completed: actual executor5765 exit1, native Apple M3 Pro, one test/zero skips. Patch04207a/q4kf7c3/full36122/testd682 unchanged across all1702 source entries. Owned group12273 and recorded descendants absent at terminal; canonical lease released; no retry.

Direct result on this exact P4/H256 fixture:
- All1024 codes match exactly; all32 scale bit patterns now match exactly (previous diagnostic had three scale mismatches). Common code SHA256cdfecda26215f8bef90149c1afb481689f2723facdbbeb7237a0f2f592e7e588; common scale SHA25600f8cb88a3a83b03c56be3fffbd306054c226dc350491462a33381bd081e04fb.
- All four normalized rows remain exact. Token7 layer0 mixed-QKV H and D now match: SHA256f2b0f3b38465ff8a36ce4ede53656d3c242f5860c95225461efc8285c5d19a3e.
- H-versus-S remains different. Serial S remains SHA256cf28fa88ba3dbc73633df3c321ed8127753dfb02f84d31f300c02b0786f7deb0; channel0 H/D bf7e14ef versus S bf7e14f0.
- Lifecycle rejection and owned-copy checks pass. Separate diagnostic1152bytes; production one read20480bytes/upload6272bytes/four encoders, total observed21632readbytes.
- Full strict overall parity remains RED for logits/GDN/KV/hidden/continuation. No oracle or tolerance changed. The whole-quantizer compilation change supports only this fixture; it does not identify a sole compiler transformation or establish general exactness/performance.

Build emitted the expected fastMathEnabled deprecation warning and duplicate-lobjc linker warning; initialization and native execution succeeded. Test1.27s/package1.794s and supervised61.480s are operational records only. Independently observed peer landing PID13018 started12:11:31 PDT, overlapping the candidate window12:10:52–12:11:53 and test12:11:51.546832–12:11:52.813724. Coordinator also reported build13166; that PID was absent at post-run census, so its exact start is not independently witnessed. Peer untouched; no performance timing claim.

Preserved artifacts: managed `fak-worker-wt-metalgemm-fa700762879c/_scratch/issue-12919-hil/{receipt.json,supervision.json,test.jsonl,test.log,before-check.json,after-check.json}`. Raw JSONL SHA256ffa347447a0a3742e055919bea9810807a395425220b6a42f1044579fc72b24b. Receipt explicitly records competing work; source candidate remains frozen and unlanded.

Next: independent source-only scoping of Q8 graph projection reduction/accumulation against native grouped GEMV, preserving P4 batching and the original strict oracle. No additional run or kernel/algorithm edit before candidate review. Parent #12349 remains open; #12853 and #12914 evidence remains provenance-distinct.

**Comment by anthony-chaudhary** [comment]
One granted functional/bitwise invocation completed for #12920: actual executor8090 exit1, native Apple M3 Pro, two serialized packages, seven named tests run, zero skips. Six pass; original full strict parity remains RED.

PASS:
- TestProjectionGraphQ8P4MatchesSerialGEMV: all four rows/all37 outputs at in1280 (40 blocks), both graph entrypoints bitwise equal to native P1.
- TestProjectionGraphMixedQuantizedSingleFence (non-P4 graph).
- TestIssue8833Q8CallerOwnedBridge.
- TestMixedQ4KQ8Observation.
- TestMetalQ8GemvMatchesCPU (including5120x5120).
- TestMetalQ8GemvGroupMatchesSingle.

Original TestMetalMTPP4StrictSerialParity executes and remains FAIL. Its direct intermediate observations now show all1024 codes and all32 scale bit patterns exact, all four norm rows exact, and H=D=S for token7 layer0 mixed QKV, common SHA256cf28fa88ba3dbc73633df3c321ed8127753dfb02f84d31f300c02b0786f7deb0. Lifecycle checks and separate1152-byte diagnostic accounting remain valid; production20480read/6272upload/four encoders unchanged.

Remaining strict mismatch logs: verification logits rows1 and3, committed boundary, layer3 KRaw/KRoPE/V, and target-hidden history at commit/continuation. Earlier layer0–2 GDN and continuation-logit mismatch logs no longer appear. This does not establish full model parity; frozen source does not contain the separately owned hidden-capture overlays.

All1702 source entries remain unchanged at fullmanifestab896f4bd0b5051b607da2d7f578cef57fd9b5c7555e53af863e6f2be8fdccf8. Owned PID/PGID69547 and recorded descendants absent at terminal; no TERM/KILL/timeout/device error. Lease released. Resident80431 retains its original incarnation and was untouched. Functional contention policy allowed concurrent CPU builders and possiblyactive resident80431; elapsed204.527s is operational only, with no performance inference.

Preserved artifacts in managed fak-worker-wt-metalgemm-be633f987f73/_scratch/issue-12920-hil/{receipt.json,test.jsonl,test.log,supervision.json,before-check.json,after-check.json}. Receipt SHA256304d3b2a359ea48ab1a24cc398d07acaeb4276f782971d2525127677cfc062c6; rawJSONL4fea76bdc9fb293cb301d240a8b22540e89f905df389e2eb4ee98ca9b8ae899e. Source remains frozen/unlanded, one invocation only, no retry.

Next: independent raw-evidence review, then scope the remaining layer3 attention/logit discrepancy separately from the already tracked hidden-history work. Parent #12349 remains open; no new edit or run is implied.

**Comment by anthony-chaudhary** [comment]
The single authorized #12927 attempt ended INCONCLUSIVE before HIL: executor99016 / owned PID+PGID39836 exited1 after154.173s. Go compilation failed writing model `$WORK/b123/_pkg_.a` and ctxmmu `$WORK/b200/_pkg_.a`: `no space left on device`. There were zero named test run events and no new arithmetic or hardware result.

The coordinator requested a capacity abort as shared disk reached226MiB free. The immediate owned-identity check found the process already terminal, so no signals were sent. Final census found no owned group or tracked descendants. Exact one-file lease released; all1702 frozen source files unchanged (manifest f6ed5e7a, patch1ed51da8, testb1009984). No retry, peer control, source deletion or cache deletion occurred. The exact run GOTMPDIR was subsequently empty; no leftover cleanup was needed.

Captured receipt SHA256 `9170ad52c586dd1df5d1b51fd8c8fb6c1dc18707abb66425efcdd03221236fc7`; raw JSONL `89da20b01fbc8a34c5baaf94ce8cbc24169421673f847eebd877b1384fdd8902`; supervision `04d64ceb0e27d3aea21def71ecf37541b8bf7188c641295897c58ff25ecfee81`. Independent foreign test/server activity was explicitly allowed for functional-only execution and was preserved. No performance claim.

Next: recover disk capacity through separately owned work, then obtain a new explicit bounded attempt grant for the unchanged reviewed candidate. #12927 and parent #12349 remain open; this attempt adds no parity evidence.

**Comment by anthony-chaudhary** [comment]
Second separately authorized attempt completed the physical diagnostic on Apple M3 Pro, engine fak-native. Executor59207 / PID+PGID56795 exited1 after74.571s: one strict test ran and failed, zero skips. No resource/device/timeout failure, capacity abort or signal. The first ENOSPC attempt remains preserved separately.

Diagnostic validity: all four replay rows match original serial.Step KRaw/KPost/V bits (12 comparisons of128 elements). Graph captured its actual incoming layer3 residual before mutation, with one command/wait/packed read of seven unique buffers totaling24576 bytes;64864 upload bytes,60 encoders,zero intermediate waits/readbacks; V alias counted once. Serial route recorded48 native events,zero fallback,12 resident GDN acceptances; CPU orchestration retained the concrete sessionQ4KKernel. QKNorm=false; Q/K weights areQ8 andV isQ4_K.

Mismatch counts by row (tokens3,5,7,11):

| Boundary | Row0 | Row1 | Row2 | Row3 |
|---|---:|---:|---:|---:|
| Incoming residual, width256 | 0 | 0 | 0 | 213 |
| Input norm, width256 | 0 | 235 | 0 | 228 |
| PackedQ, width512 | 0 | 380 | 0 | 355 |
| KRaw, width128 | 0 | 97 | 0 | 97 |
| V, width128 | 0 | 109 | 0 | 115 |
| KPost, width128 | 31 | 105 | 32 | 100 |
| Host RoPE on graph KRaw versus graph KPost | 31 | 33 | 32 | 30 |

Causal scope: row3 already differs before layer3, requiring earlier-layer localization. Row1 has identical incoming input but differing norm (firstchannel0 serialbf7f9a44 versus graphbf7f9a46), directly identifying a norm boundary difference. Rows0/2 remain exact throughKRaw but differ atKPost; replaying host RoPE on the actual graphKRaw differs on all rows, separating postKRaw arithmetic from upstream errors. These observations do not establish a Q4/Q6 projection defect or select a production fix.

The original strict logits rows1/3, committed boundary and layer3 caches still fail. Layer0 H=D=S remains exactcf28fa88 and all1024 codes/all32scale bits match. Hidden bookkeeping is now explicit: serial0rows/0tokens; target36rows/36tokens,32nilprefix rows and suffix[3,5,7,11] aftercommit/continuation. This is not evidence of losing the verified suffix.

Receipt SHA256 `00fe57e8d3c82c1a86fc5e660b9bbcd18499627eb683952aeeb1b1354881c627`; rawJSONL `112ecbb29471d34f378ea86615a9b9cbea9e58cc08432d5a227735d0755ee29c`; supervision `5e38392d950bd01d7ef6bb78afdb5022830c248b3506331ecdaef31a1eb76a08`. All1702 source files unchanged (manifestf6ed5e7a/testb1009984/patch1ed51da8). Fourteen capacity samples stayed above2GiB (minimum6.86GiB). Owned group/descendants absent; exact lease released. No timing/performance claim amid permitted peer overlap.

Next: independently validate causal interpretation and scope the smallest production arithmetic experiment. No further run or source edit is authorized yet; parent #12349 remains open.

**Comment by anthony-chaudhary** [comment]
## Resolved by `0cb93174c` — independently witnessed

Independent verification in worktree `ticket-12349` confirms the issue outcome is already landed and green.

**Landing:** `0cb93174cca505cee6dc44cb497eb45d3e09281b` "perf(model): route Metal hybrid MTP through wide-M batched verification (#12349) (fak model)" touched exactly the routed files (`internal/model/verify.go`, `internal/model/metal_mtp.go`, `internal/model/metal_mtp_test.go`). `dos commit-audit` → `OK` / `diff-witnessed`. In HEAD ancestry.

**Witness (issue command):** `go test ./internal/model -run '^TestMetalMTPWideMUsesBatchedVerifier$' -count=1` → PASS (4/4 subtests):
- `admitted_hybrid_envelope_selects_batched_dispatch_with_serial_parity` — admission + `OneOperation=true`, `Path=batched-target-verify-v1`, serial-parity logits/state.
- `metal_flag_admitted_in_hybrid_envelope` — `Metal=true` admitted on the generic envelope.
- `verify_forward_batched_matches_serial_step_logits_and_cache` — bit-exact logits + cache pos.
- `unsupported_shapes_fallback_to_typed_fak_native_serial_decode` — F16 stays serial (`OneOperation=false`, `TargetDecodeSteps=2`, `Path=ordinary-target-decode-v1`).

**Production tuple (real Apple Silicon Metal Q4_K Qwen3.8 hybrid):** reaches wide-M verification via the transaction seam `metal_mtp.go:871-872` → `stepRoundQwen35P4Locked` → `specbind.go:143` → `qwen35MTPMetalP4Verify` → `Qwen35MetalMTPVerifyPanel` → `encodeQwen35MetalMTPP4`, i.e. before the generic `verifyForwardBatchedOK` branch. The generic f32/hybrid admission (`isMetalQwenHybridVerifyEnvelope`) is correctly narrow (rejects `MetalQ4K`, `Q4K`, `Quant`, `F16`, `Backend`, `PrecisionPolicy`).

**Acceptance:** admit only the supported envelope ✅; batched dispatch selected with serial-equivalent logits/state ✅; unsupported shapes stay typed fak-native serial, no engine switch ✅.

**Owned-path validation:** `go vet ./internal/model` clean; `gofmt -l` clean on all three files.

**Nonblocking follow-on:** `#13079` removes an unreachable `s.MetalQ4K` branch in `verifyForwardQwenHybrid` (`verify.go:67`) found during review — out of this issue's routed scope, filed separately.

No new code change was needed; closing as resolved by the cited commit. Sibling HIL strict-parity investigations (#12914/#12919/#12920/#12927) remain separate and do not gate this routed admission issue.
